### PR TITLE
CAS-03F: Register manifest ownership by storage pool

### PR DESCRIPTION
**Pull request by ScottArbeit**
# Pull Request

## Linked issue

Related to #442.

Part of #426 and #424.

This PR intentionally targets `epic/426-storagepool-routing`, not `main`, not `epic/424-storagepool-cas`, and not
`epic/343-blake3-sha256-version-hashes`. The sub-issue will be closed manually only after the reviewed slice is merged
into the mini-integration branch.

## Summary

Implements CAS-03F by registering and removing manifest ownership with explicit `StoragePoolId` identity instead of
address-only or current-route assumptions.

Implemented behavior:

- Repository manifest ownership identity includes `RepositoryId`, `StoragePoolId`, and `ManifestAddress`.
- Manifest contribution identity includes `StoragePoolId` plus `ContentBlockAddress`, preserving same-address
  cross-pool separation.
- Save, Commit, Checkpoint, child-directory, and physical deletion/expiry paths use stored manifest pool evidence.
- Upload-session/SDK/storage finalization propagation now preserves durable manifest `StoragePoolId` evidence needed by
  later reference ownership.
- Whole-file-only paths remain unaffected and do not require manifest-pool evidence.

## Touched paths

Core implementation:

- `src/Grace.Types/Common.Types.fs`
- `src/Grace.Types/RepositoryContentCounter.Types.fs`
- `src/Grace.Types/ManifestContributionWorkflow.Types.fs`
- `src/Grace.Actors/RepositoryContentCounter.Actor.fs`
- `src/Grace.Actors/ManifestContributionWorkflow.Actor.fs`
- `src/Grace.Actors/Reference.Actor.fs`
- `src/Grace.Actors/DirectoryVersion.Actor.fs`
- `src/Grace.Actors/UploadSession.Actor.fs`
- `src/Grace.Server/Storage.Server.fs`
- `src/Grace.SDK/ManifestUpload.SDK.fs`

Tests and evidence:

- `src/Grace.Server.Unit.Tests/RepositoryContentCounter.Actor.Tests.fs`
- `src/Grace.Server.Unit.Tests/ManifestContributionWorkflow.Actor.Tests.fs`
- `src/Grace.Server.Unit.Tests/SaveBoundary.Actor.Tests.fs`
- `src/Grace.Server.Unit.Tests/UploadSession.Actor.Tests.fs`
- `src/Grace.Server.Unit.Tests/Storage.Server.Tests.fs`
- `src/OpenAPI/Storage.Components.OpenAPI.yaml`
- `src/OpenAPI/Grace.OpenAPI.yaml`
- `src/OpenAPI/Grace.OpenAPI.3.1.2.yaml`
- `src/OpenAPI/OpenAPI.ProofManifest.json`
- `sdk/generated/matrix/generator-matrix-evidence.json`
- `src/Grace.Actors/AGENTS.md`

## Owned-path compliance

- [x] Changed paths match the linked issue's owned paths and explicitly justified path expansion.
- [x] The worker recorded path expansion for `UploadSession.Actor.fs`, `Storage.Server.fs`, and `ManifestUpload.SDK.fs`
  as contract propagation for durable manifest `StoragePoolId` evidence.
- [x] This PR does not implement #441 shared CAS download authorization.
- [x] This PR does not implement #443 annotation materialization.

## Validation run

Implementation worker validation for `754bba16115975e0c05b724a1889f7e6998e9226`:

- Focused actor/save-boundary tests: passed, 54/54.
- OpenAPI freshness proof: passed for this slice; `prove-openapi` still reports unrelated existing
  quality/package/protocol pending gates.
- Targeted Fantomas: passed.
- MarkdownLint for touched Markdown: passed.
- `pwsh ./scripts/validate.ps1 -Fast`: passed after stale manifest `StoragePoolId` test setup was fixed.
- `git diff --check`: passed.

### CI Fix 2 - 2026-06-21

- Fix worker: GPT-5.5 Medium worker `019eea8b-969c-7c01-981f-92bab77fb648`.
- Fix commit: `b9d5f6e370b4eac974c70d0fd475503336c73383` (`Fix manifest ownership CI regressions`).
- CI failure source: GitHub `Validate / full` run `27906933849` failed three `Grace.Server.Tests` integration cases on
  `ff75d97900599440d9220819385459f8b4879ede`.
- Root causes: the child-directory root-rejection regression used short child hashes that could collide with unrelated
  shared fixture roots; two manifest download integration flows omitted the now-required manifest `StoragePoolId` and
  therefore exercised the wrong failure path or failed closed.
- Fix summary: branch child-directory integration tests now use the full child SHA/BLAKE3 hashes, and manifest download
  integration requests now set `downloadUriParameters.StoragePoolId <- manifest.StoragePoolId`.
- Validation: targeted Fantomas passed unchanged; Release build for `Grace.Server.Tests` passed; focused three-test
  `Grace.Server.Tests` filter passed 3/3; `git diff --check` passed. `validate -Fast` was skipped because it excludes
  `Grace.Server.Tests` and would not validate this integration-test-only fix.
- Prevention: integration fixture and contract-call drift; integration tests now use non-colliding directory-version
  hashes and follow the manifest download contract's recorded storage-pool evidence requirement.
- Final gate: GitHub `Validate / full` run `27907320916` passed on `b9d5f6e370b4eac974c70d0fd475503336c73383`; Codex Code Review Bot returned 👍 at `2026-06-21T14:31:52Z`; PR #449 merged into `epic/426-storagepool-routing` as `df68d3273f5fb334d24eb6aa1608644f211081f9`.`r`n### Review Fix 2 - 2026-06-21

- Fix worker: GPT-5.5 Medium worker `019eea72-bacb-7473-a85e-5b96247ecf99`.
- Fix commit: `ff75d97900599440d9220819385459f8b4879ede` (`Fix manifest pool authorization boundaries`).
- Codex review source: review on `05252094868a8c7316e22802e241586756ab2b81`.
- Save-boundary pool authorization: manifest save-boundary validation now proves the manifest block through scoped
  finalized metadata for `(StoragePoolId, RepositoryId, authorized scope, ManifestAddress, ContentBlockAddress)` before
  trusting the recorded manifest pool for ownership planning.
- Manifest download routing: manifest downloads now carry the durable `FileManifest.StoragePoolId`; the server authorizes
  that recorded pool against finalized scoped metadata and returns the stored placement instead of re-resolving the
  repository's current route.
- Contract updates: `GetContentBlockDownloadUriParameters` now includes `StoragePoolId`, with OpenAPI projections and
  generated TypeScript, Python, Rust, and SDK harness metadata refreshed.
- Validation: targeted Fantomas passed; focused Server.Unit build and regression filter passed 54/54; focused CLI build
  and `ManifestDownloadSdkTests` passed 8/8; OpenAPI projection passed; SDK client freshness check passed; generator
  matrix passed; `prove-openapi` freshness/projection/matrix checks passed with existing unrelated pending gates;
  `git diff --check` passed; `pwsh ./scripts/validate.ps1 -Fast` passed.
- Prevention: authorization and route-evidence gap; #442 now gates save-boundary ownership on scoped finalized manifest
  evidence and gates manifest downloads on the manifest's recorded storage-pool evidence.
### Review Fix - 2026-06-21

- Fix worker: GPT-5.5 Medium worker `019ee8f3-fcac-76a2-b4cf-e1f4ca25246f`.
- Fix commit: `05252094868a8c7316e22802e241586756ab2b81` (`Fix manifest storage-pool review findings`).
- Codex review source: review on `754bba16115975e0c05b724a1889f7e6998e9226`.
- Direct finalize callers: resolved by current-head evidence plus CI-fix commit
  `ec654907beac49e98a2cd29a52ee929eea5d686d`; direct HTTP integration helpers now finalize manifests with the
  upload session's resolved `StoragePoolId`, SDK finalize already rewrites to the session pool, and the runtime guard
  remains intact.
- Workflow range pool validation: added fail-closed validation that rejects range identities whose `StoragePoolId` does
  not match the manifest contribution workflow `StoragePoolId`.
- Generated SDK models: refreshed the OpenAPI generator matrix and SDK harness metadata so tracked generated
  TypeScript, Python, Rust, and harness outputs include `FileManifest.StoragePoolId`.
- Validation: `pwsh ./sdk/scripts/invoke-generator-matrix.ps1` passed;
  `pwsh ./sdk/scripts/generate-sdk-clients.ps1 -Mode Check` passed after regeneration; targeted Fantomas passed;
  focused `ManifestContributionWorkflowActorTests` passed 14/14; `pwsh ./scripts/validate.ps1 -Fast` passed;
  `git diff --check` passed.
- Prevention: contract-propagation and negative-proof gap; #442 now carries storage-pool identity through generated
  contracts and rejects cross-pool contribution ranges before active manifest counts can be adjusted under the wrong
  workflow identity.

### CI Fix - 2026-06-21

- Fix worker: GPT-5.5 Medium worker `019ee8e9-b1ec-7673-8508-d481b8880800`.
- Fix commit: `ec654907beac49e98a2cd29a52ee929eea5d686d` (`Fix manifest storage pool test setup`).
- CI failure source: `Validate / full` run `27896075510` failed six `Grace.Server.Tests` integration cases because test
  helpers finalized manifests with `StoragePoolId = default` while upload sessions used generated storage pools.
- Fix summary: server integration test finalization paths now rebuild `FileManifest` values from the upload session's
  resolved `StoragePoolId` and recompute the manifest address from the exact finalized manifest record.
- Validation: targeted Fantomas unchanged;
  `dotnet build --configuration Release src/Grace.Server.Tests/Grace.Server.Tests.fsproj` passed; focused failing
  integration-test filter passed 6/6; `git diff --check` passed.
- GitHub validation: `Validate / full` run `27896436173` passed on
  `ec654907beac49e98a2cd29a52ee929eea5d686d`.
- Prevention: test-evidence gap; integration helpers now preserve the same upload-session storage-pool evidence required
  by the production finalize guard, rather than weakening the guard.

### Implementation - 2026-06-21

- Implementation worker: GPT-5.5 Medium worker `019ee8cd-34f9-7071-a5db-fa7d191da7a1`.
- Initial implementation commit: `754bba16115975e0c05b724a1889f7e6998e9226`.
- Worker bot-prevention self-review: completed for correctness/edge cases, tests, stale identity assumptions,
  replay/idempotency, deletion symmetry, generated artifacts, and validation gaps.
- Initial validation: focused actor/save-boundary tests passed 54/54; OpenAPI freshness proof passed for this slice;
  targeted Fantomas passed; MarkdownLint for touched Markdown passed; `pwsh ./scripts/validate.ps1 -Fast` passed after
  stale manifest `StoragePoolId` test setup was fixed; `git diff --check` passed.

## Docs impact

Updated `src/Grace.Actors/AGENTS.md` for actor-local StoragePool manifest ownership guidance.

## Residual risk

- OpenAPI proof still reports pre-existing unrelated quality/package/protocol pending gates; freshness and matrix hash
  checks for this slice pass.
- #441 owns shared CAS download authorization that consumes this ownership proof.
- #443 owns annotation materialization from stored manifest CAS placement.
- #444 owns the final #426 generated-contract and integration audit.

### CODE REVIEW

File: src/Grace.Actors/UploadSession.Actor.fs [hunk 1]
```diff
@@ -946,6 +946,14 @@ module UploadSession =
                     correlationId
                     $"FileManifest ChunkingSuiteId must match UploadSession ChunkingSuiteId. Expected {session.ChunkingSuiteId}, actual {manifest.ChunkingSuiteId}."
             )
+        elif String.IsNullOrWhiteSpace manifest.StoragePoolId then
+            Error(graceError correlationId "FileManifest StoragePoolId must be recorded before finalization.")
+        elif manifest.StoragePoolId <> session.StoragePoolId then
```
[the remark this answers is not in the document]
Comment (ScottArbeit) [reply src/Grace.Actors/UploadSession.Actor.fs:951]: ## Review/Fix: Direct finalize callers use the session pool

**Status:** Fixed across `ec654907beac49e98a2cd29a52ee929eea5d686d` and
`05252094868a8c7316e22802e241586756ab2b81`.

**Validation:** GitHub `Validate / full` passed on `ec654907beac49e98a2cd29a52ee929eea5d686d`; the review-fix worker
also passed generator matrix, SDK freshness, targeted Fantomas, focused `ManifestContributionWorkflowActorTests` 14/14,
`pwsh ./scripts/validate.ps1 -Fast`, and `git diff --check` on the current head.

**Prevention:** Contract-propagation gap; direct finalize test callers and generated clients now preserve explicit
manifest `StoragePoolId` evidence instead of weakening the production finalize guard.

### Summary

The known non-SDK direct HTTP finalization callers in the full storage integration helpers now rebuild the manifest from
the upload session's resolved `StoragePoolId` before finalization and recompute the address from that exact manifest. The
SDK path already rewrites finalize manifests to the session pool, and the review-fix commit refreshed generated SDK
outputs so tracked clients carry `FileManifest.StoragePoolId` as part of the contract.

File: src/Grace.Actors/ManifestContributionWorkflow.Actor.fs [hunk 2]
```diff
@@ -133,11 +137,13 @@ module ManifestContributionWorkflow =
         =
         if start.RepositoryId = RepositoryId.Empty then
             Some(graceError metadata.CorrelationId "ManifestContributionWorkflow requires a non-empty RepositoryId.")
+        elif String.IsNullOrWhiteSpace start.StoragePoolId then
+            Some(graceError metadata.CorrelationId "ManifestContributionWorkflow requires a non-empty StoragePoolId.")
```
[the remark this answers is not in the document]
Comment (ScottArbeit) [reply src/Grace.Actors/ManifestContributionWorkflow.Actor.fs:152]: ## Review/Fix: Reject workflow ranges from other storage pools

**Status:** Fixed in `05252094868a8c7316e22802e241586756ab2b81`.

**Validation:** Focused `ManifestContributionWorkflowActorTests` passed 14/14; targeted Fantomas passed;
`pwsh ./scripts/validate.ps1 -Fast` passed; `git diff --check` passed.

**Prevention:** Negative-proof gap; the workflow now validates range storage-pool identity before active manifest counts
can be adjusted under the wrong workflow identity.

### Summary

`ManifestContributionWorkflow.Actor.fs` now fails closed when a `Start` command includes any range whose
`StoragePoolId` does not match `start.StoragePoolId`. The fix adds regression coverage for a cross-pool range so future
callers cannot mix pool B ranges into a workflow keyed by pool A.

File: src/OpenAPI/Grace.OpenAPI.3.1.2.yaml [hunk 3]
```diff
@@ -4572,6 +4572,8 @@ components:
           $ref: '#/components/schemas/ChunkingSuiteId'
         FileContentHash:
           $ref: '#/components/schemas/FileContentHash'
+        StoragePoolId:
+          $ref: '#/components/schemas/StoragePoolId'
```
[the remark this answers is not in the document]
Comment (ScottArbeit) [reply src/OpenAPI/Grace.OpenAPI.3.1.2.yaml:4578]: ## Review/Fix: Regenerate checked-in SDK models for StoragePoolId

**Status:** Fixed in `05252094868a8c7316e22802e241586756ab2b81`.

**Validation:** `pwsh ./sdk/scripts/invoke-generator-matrix.ps1` passed;
`pwsh ./sdk/scripts/generate-sdk-clients.ps1 -Mode Check` passed after regeneration;
`pwsh ./scripts/validate.ps1 -Fast` passed; `git diff --check` passed.

**Prevention:** Generated-contract freshness gap; the tracked generated matrix and SDK harness metadata are refreshed
with `FileManifest.StoragePoolId` instead of leaving clients on a stale manifest shape.

### Summary

The generator matrix and checked-in generated outputs were refreshed for TypeScript, Python, Rust, and SDK harness
metadata. The tracked `FileManifest` models now include `storagePoolId` in the model surface and serialization paths,
with updated generator evidence committed alongside the generated files.

**Comment by ScottArbeit** [comment]
@codex review

File: src/Grace.Actors/DirectoryVersion.Actor.fs [hunk 4]
```diff
@@ -485,7 +485,7 @@ module DirectoryVersion =
                 while blockIndex < manifest.Blocks.Count && error.IsNone do
                     let block = manifest.Blocks[blockIndex]
                     let query: ContentBlockRangeQuery = { OrdinalStart = 0; OrdinalCount = 1 }
-                    let! presence = getRangePresence block.Address query
+                    let! presence = getRangePresence manifest.StoragePoolId block.Address query
```
[the remark this answers is not in the document]
Comment (ScottArbeit) [reply src/Grace.Actors/DirectoryVersion.Actor.fs]: ## Review/Fix: Authorize manifest pools before save-boundary checks

**Status:** Fixed in `ff75d97900599440d9220819385459f8b4879ede`.

**Validation:** Targeted Fantomas passed; focused Server.Unit build and regression filter passed 54/54;
`pwsh ./scripts/validate.ps1 -Fast` passed; `git diff --check` passed.

**Prevention:** Authorization and route-evidence gap; save-boundary ownership planning now requires scoped finalized
manifest evidence before trusting the manifest's recorded storage pool.

### Summary

Save-boundary validation now carries each manifest with its authorized file scope and validates blocks through scoped
finalized metadata for `(StoragePoolId, RepositoryId, authorized scope, ManifestAddress, ContentBlockAddress)`. That
proves the recorded manifest pool is finalized for the saving repository/path scope before the later counter and
contribution-workflow ownership path uses it.

File: src/Grace.Types/Common.Types.fs [hunk 5]
```diff
@@ -331,6 +331,8 @@ module Common =
             ChunkingSuiteId: ChunkingSuiteId
             [<Key(5)>]
             FileContentHash: FileContentHash
+            [<Key(6)>]
+            StoragePoolId: StoragePoolId
```
[the remark this answers is not in the document]
Comment (ScottArbeit) [reply src/Grace.Types/Common.Types.fs:335]: ## Review/Fix: Route manifest downloads through the recorded pool

**Status:** Fixed in `ff75d97900599440d9220819385459f8b4879ede`.

**Validation:** Focused CLI build and `ManifestDownloadSdkTests` passed 8/8; OpenAPI projection passed; SDK client
freshness check passed; generator matrix passed; `pwsh ./scripts/validate.ps1 -Fast` passed; `git diff --check` passed.

**Prevention:** Contract-propagation and route-drift gap; manifest downloads now send and authorize the manifest's
durable `StoragePoolId` instead of recomputing the repository's current route.

### Summary

`GetContentBlockDownloadUriParameters` now includes `StoragePoolId`, and `ManifestDownload` supplies it from the durable
`FileManifest`. `Storage.GetContentBlockDownloadUri` authorizes that recorded pool against finalized scoped metadata and
uses the stored placement for the download URI. OpenAPI projections and generated SDK artifacts were refreshed for the
new request shape.
